### PR TITLE
Add `flow: wrap` option to handle circular axes 

### PR DESCRIPTION
**Pull request by dsavoiu**
This PR implements a new `wrap` option for handling over-/underflow values on circular axes. A typical use case is for corrections that use the azimuthal angle *φ* as in input, which is unique up to integer multiples of 2π. This should alleviate issues with inputs landing outside the axis interval due to limited float precision, in which case `flow: error` would unnecessarily break workflows and the other existing flow behaviors like `clamp` or a fixed value may give slightly wrong results.

The implementation is analogous to the bin lookup on circular axes, as implemented in *boost-histogram* (see [code](https://github.com/boostorg/histogram/blob/66842660c0bad0ce80c6db5fe68c7d7e4ec00be4/include/boost/histogram/axis/regular.hpp#L310-L316)).

Happy to hear your feedback!

### CODE REVIEW

File: src/correction.cc [hunk 1]
```diff
@@ -467,6 +486,9 @@ Binning::Binning(const JSONObject& json, const Correction& context)
   else if ( flowbehavior == "error" ) {
     flow_ = _FlowBehavior::error;
   }
+  else if ( flowbehavior == "wrap" ) {
```
Comment (nsmith-) [inline src/correction.cc:489]: Given the number of options here, and the duplicated code in `MultiBinning::MultiBinning` can you factor this out into a free function `parse_flowbehavior`?

File: src/correction.cc [hunk 2]
```diff
@@ -172,6 +185,9 @@ namespace {
       else if ( flow == _FlowBehavior::error ) {
         throw std::runtime_error("Index below bounds in "s + name + " for input argument " + std::to_string(variableIdx) + " value: " + std::to_string(value));
       }
+      else if ( flow == _FlowBehavior::wrap ) {
+        throw std::logic_error("I should not have ever seen an underflow");
```
Comment (nsmith-) [inline src/correction.cc:189]: Is this always true? I suppose so if `norm_value >= 0` is guaranteed.

File: src/correction.cc [hunk 3]
```diff
@@ -183,6 +199,9 @@ namespace {
       else if ( flow == _FlowBehavior::error ) {
         throw std::runtime_error("Index above bounds in "s + name + " for input argument " + std::to_string(variableIdx) + " value: " + std::to_string(value));
       }
+      else if ( flow == _FlowBehavior::wrap ) {
+        throw std::logic_error("I should not have ever seen an overflow");
```
Comment (nsmith-) [inline src/correction.cc:203]: Here I suppose we are safe as long as `norm_value < 1` is guaranteed and `low + (1 - eps) * (high - low) < high` as well. The latter I might worry about.

**Review by nsmith- — CHANGES_REQUESTED** [review]
Overall looks good. There are some floating point order of operations concerns but you guard against them so it is at least failing safe in the event we run into a corner case.

**Comment by nsmith-** [comment]
@dsavoiu did you have a chance to look at this yet?

File: src/correction.cc (hunk 1 above)
Comment (dsavoiu) [reply src/correction.cc]: Sure, I added a commit with that code factored out to a separate function.

File: src/correction.cc (hunk 3 above)
Comment (dsavoiu) [reply src/correction.cc:219]: For the overflow there are indeed some rare cases where these expressions do not hold.

I checked this for a few hand-picked values, using `float` instead of `double` for simplicity and looping through various combinations for (`low`, `high`, `input_value`) to see if any would trigger the overflow condition. I explicitly included numbers here that are either very close together like `±(1.0f ± FLT_EPSILON)` or very far apart like `±FLT_MAX`, since these seemed the most likely to generate these edge cases.

It turns out that `norm_value == 1` can occur when an `input_value` is provided such that `(input_value - low) / (high - low) > -eps`, i.e. the input value is slightly less than `low` but the difference is too small to be resolved with machine precision with respect to the interval width. Logically, this situation should probably be treated as one would any other value below the `low` threshold, meaning that the lookup should wrap around the circular axis and the content of the highest bin should be returned.

A different edge case seems to occur when `high` and `low` are either very close together (on the order of `eps` or below) or very far apart (e.g. both near `±FLT_MAX`). In these cases, the condition `low + (1 - eps) * (high - low) < high` may no longer hold and the `norm_value` can also become nonfinite as a result (i.e. `-nan`), although it it also possible to obtain finite values in `[0, 1)` depending on the `input_value`. Overall, it seems unlikely that these situations would occur in practice, so while one should be aware that they can happen it's probably not necessary to handle them explicitly (other than maybe adding a few safeguards against `nan` values).

There may also be other pathological scenarios that I did not catch. From what I can tell, the most relevant change here would be to catch the `norm_value == 1` case and manually return the corresponding bin, but I'm happy to hear other opinions and suggestions.

**Comment by dsavoiu** [comment]
Hi @nsmith-, sorry, got a bit sidetracked on this. I factored out the flow behavior parsing to a separate function like you asked. I also did some testing to see if machine precision would affect the under/overflow behavior. You are correct that for some edge cases (details inline above), the conditions you provided no longer hold and the overflow exception handling is triggered.

However, as this seems to happen only for some pathological cases where the upper and lower edges are many orders of magnitude apart (which is unlikely to occur in practice), and since they are caught as exceptions, I would say the danger of running into these unnoticed is minimal. But do please let me know if you think this needs more checks or a more detailed look.

File: src/correction.cc [hunk 4]
```diff
@@ -149,20 +165,33 @@ namespace {
...
             throw std::runtime_error(std::move(msg));
         }
       }
 
-      std::size_t binIdx = bins->n * ((value - bins->low) / (bins->high - bins->low));
+      double norm_value = ((value - bins->low) / (bins->high - bins->low));
+      if (flow == _FlowBehavior::wrap) {
+        norm_value -= std::floor(norm_value);
+      }
+      std::size_t binIdx = bins->n * norm_value;
```
Comment (nsmith-) [inline src/correction.cc:181]: What if, rather than using `std::floor` to wrap, we use `binIdx = binIdx % bins->n` for `_FlowBehavior::wrap`?

File: src/correction.cc (hunk 3 above)
Comment (nsmith-) [reply src/correction.cc:219]: Thanks for looking into this! I think it's a reasonable conclusion. See my other comment, and if you agree maybe that is a way out of this pickle.

File: src/correction.cc (hunk 4 above)
Comment (dsavoiu) [reply src/correction.cc:181]: Thanks for the suggestion! The major limitation with this is that `norm_value` can also be negative, so the cast from signed `double` to the integer `std::size_t` in `std::size_t binIdx = bins->n * norm_value` would lead to undefined behavior. For positive values is mostly works, but there are some caveats there as well, for example if `bins->n * norm_value` overflows past `ULONG_MAX`.

Another thing to consider is that a calculation based on modulo arithmetic would anyway only apply to uniform binning, whereas the normalize-and-floor approach can also work for nonequidistant bin edges.

**Review by nsmith- — APPROVED** [review]
LGTM
